### ISSUE TITLE
Non-recursive transform functions

### ISSUE THREAD

**Issue by metasoarous** [Engelberg/instaparse#88]
The transform functions as currently implemented are recursive, and could blow the stack for large trees. Any chance of implementing using zippers?

**Comment by Engelberg** [comment]
Initial thoughts:
1. It's not a problem for all large trees, just extremely *deep* trees.
2. Nearly all of Clojure's functions struggle with deep trees, so if you have a tree that is deep enough to cause problems with transform, you'll probably run into problems elsewhere when you go to manipulate and process those trees.  Initially, I tried to make instaparse insensitive to tree depth, but extremely deep trees inevitably would blow up other parts of my code, so eventually I gave up and just embraced this as a limitation of Clojure and coded instaparse accordingly.  So changing transform alone isn't going to solve the problem -- there are other aspects of instaparse that use the stack to process the trees.
3. I haven't studied the implementation of zippers, but are you sure that zippers don't use stack space?  I'm not so sure using zippers would solve the problem.
4. Generally, you can make the tree flatter with an appropriate grammar.  For example, instead of using S = 'a' | 'a' S, use S = 'a'+.

That said, if you're having a problem with blowing the stack, and you think that implementing transform with zippers is a solution, I'd be curious to see it and to compare the performance.

**Comment by metasoarous** [comment]
Here's my context: I'm working on some phylogenetic extensions to [clj-biosequence](https://github.com/s312569/clj-biosequence), in particular a [Newick](http://en.wikipedia.org/wiki/Newick_format) parser. While most applications of phylogenetic trees would be totally within the realm of recursion, I've worked in metagenomics and have seen some pretty big (and deep) trees there. However, it's likely I'm being overly cautious; I should dig up some of the old data I worked on, see how big those trees are and try parsing/transforming them.

In any case, I hear you that there are other parts of instaparse (and Clojure) that don't handle deep trees very well, and that this may a lost cause if it's an issue at all. I'll just have to tinker a bit to see what the limits are. I _am_ in my implementation trying to be as flat as I can though, so that should be a help.

As for zippers, I _think_ they aren't stack consuming, but I could be wrong. I seem to remember them coming up in a post somewhere about avoiding stack overflows, but I can't seem to find that now so maybe I misinterpreted something.

Thanks

PS By the way, I was able to implement the guts of the parser almost entirely by plopping in the grammer definition from Wikipedia (sans some flattness modifications). You've built a wonderful thing :-)
